### PR TITLE
Add support for custom User Agent

### PR DESCRIPTION
**Pull request by ryanccn**
Allows overriding the User Agent for no specific purpose

### CODE REVIEW

File: launcher/ui/pages/global/APIPage.ui [hunk 1]
```diff
@@ -220,6 +220,31 @@
...
+          <string>User Agent</string>
+         </property>
+         <layout class="QVBoxLayout" name="verticalLayout_7">
+          <item>
+           <widget class="QLineEdit" name="userAgentLineEdit"/>
+          </item>
+          <item>
+           <widget class="QLabel" name="userAgentLabel">
+            <property name="text">
+             <string>Enter a custom User Agent here. The special string ${launcher_version} will be replaced with the version of the launcher.</string>
```
Comment (flowln) [inline launcher/ui/pages/global/APIPage.ui]: maybe `$LAUNCHER_VER` would be better, to keep things similar to the ones in the custom commands page

**Review by flowln — COMMENTED** [review]
Since you're at it, why not make this page a scrollable one? We are almost reaching the vertical limit here :x

File: launcher/ui/pages/global/APIPage.ui (hunk 1 above)
Comment (ryanccn) [reply launcher/ui/pages/global/APIPage.ui]: hm okay

Comment (ryanccn) [reply launcher/ui/pages/global/APIPage.ui]: (anyways, this isn't even actually implemented)

Comment (ryanccn) [reply launcher/ui/pages/global/APIPage.ui]: now it is

**Comment by Scrumplex** [comment]
What about a tab-based approach?

Page 1
Page 2
Page 3

**Comment by Scrumplex** [comment]
I just pushed my commit here, just to have it somewhere. If you guys don't like it, just drop the commit again

**Comment by ryanccn** [comment]
> I just pushed my commit here, just to have it somewhere. If you guys don't like it, just drop the commit again

LGTM

**Review by flowln — APPROVED** [review]
looks good 
_proceeds to settings the UA to MMC's one and use their meta servers to confuse everyone on #support :pofat:_ /j

**Comment by Scrumplex** [comment]
> _proceeds to settings the UA to MMC's one and use their meta servers to confuse everyone on #support :pofat:_ /j

but our meta servers are way cooler, because they communicate supported JRE versions :sunglasses:

File: launcher/Application.cpp [hunk 2]
```diff
@@ -1553,3 +1554,23 @@ QString Application::getCurseKey()
...
+
+    return BuildConfig.USER_AGENT;
+}
+
+QString Application::getUserAgentUncached()
+{
+    QString uaOverride = m_settings->get("UserAgentOverride").toString();
+    if (!uaOverride.isEmpty()) {
+        return uaOverride.replace("$LAUNCHER_VER", BuildConfig.printableVersionString());
+    }
```
Comment (Scrumplex) [inline launcher/Application.cpp]:
```suggestion
    QString uaOverride = m_settings->get("UserAgentOverride").toString();
    if (!uaOverride.isEmpty()) {
        uaOverride += " (Uncached)";
        return uaOverride.replace("$LAUNCHER_VER", BuildConfig.printableVersionString());
    }
```

**Review by Scrumplex — APPROVED** [review]

**Comment by istudyatuni** [comment]
Can you make `$LAUNCHER_VER` copyable on the miscellaneous tab?

**Review by txtsd — APPROVED** [review]

**Comment by ryanccn** [comment]
merge wen

**Pull request merged**